File: apps/web-gallery/lib/replicate.ts
import { z } from "zod";

/**
 * For image conversions, ML models are used in Replicate
 * https://replicate.com/docs/reference/http
 * 
 * For this specific image -> GTA image conversion, we're using the gta5_lora model
 * https://replicate.com/cloneofsimo/gta5_lora
 */
const REPLICATE_BASE_URL = "https://api.replicate.com/v1/predictions";

const predictionSchema = z.object({
    completed_at: z.string().nullable(),
    created_at: z.string(),
    error: z.string().nullable(),
    id: z.string(),
    input: z.any(),
    output: z.string().array().nullable(),
    status: z.string(),
    version: z.string(),
})

type Prediction = z.infer<typeof predictionSchema>;

export const createPrediction = async (input: string): Promise<Prediction> => {
  const response = await fetch(REPLICATE_BASE_URL, {
    method: "POST",
    headers: {
      Authorization: `Token ${process.env.REPLICATE_API_TOKEN}`
    },
    body: JSON.stringify({
      version: process.env.GTA5_LORA_VERSION,
      input: {
        prompt: "a photo of <1> gtav style",
        image: input
      }
    })
  });
  const json = await response.json();
  const schema = predictionSchema.safeParse(json);
  if(!schema.success) {
    throw new Error(JSON.stringify(schema.error));
  }
  return schema.data;
}

export const getPrediction = async (id: string) => {
    const url = `${REPLICATE_BASE_URL}/${id}`;
    const response = await fetch(url, {
        headers: {
            Authorization: `Token ${process.env.REPLICATE_API_TOKEN}`
        },
    });
    const json = await response.json();
    const schema = predictionSchema.safeParse(json);
    if(!schema.success) {
      throw new Error(JSON.stringify(schema.error));
    }
    return schema.data;
}

File: apps/background-jobs/src/shared/replicate.py
import replicate
from io import BytesIO

model = replicate.models.get("cloneofsimo/gta5_lora")
version = model.versions.get(
    "3c2f37b92610fb0d92d6acdb3e7754e467790c7ed76760fcc0b952f9c0ec49cd"
)


def get_prediction(id: str):
    return replicate.predictions.get(id)


def create_prediction(image_data: BytesIO):
    return replicate.predictions.create(
        version=version,
        input={
            "prompt": "a photo of <1> gtav style",
            "image": image_data,
            "width": 1024,
            "height": 768,
        },
    )
